### DISCUSSION TITLE
How to avoid negative simulated concentration values with add+prop RUV model

### DISCUSSION THREAD

**Discussion by mark0935git** [metrumresearchgroup/mrgsolve#990] [General]
Dear Mrgsolve developers,

I tried to simulate central compartment concentration using the PopPK model below. However, the simulated concentrations are negative sometimes.

I saw one similar post from 2017, where you suggested to "use exponential error model or resimulate EPS(1) with simeps() to keep concentrations positive"

Could you please give some examples on how to implement these approches?

Many thanks in advance!

```
PopPK_IV_2CMT_LinCL_Model = "
[PARAM] @annotated
TV_CL:   0.027   : Linear CL from central compartment (L/h)
TV_V1:   2.95    : Volume of central compartment (L)
TV_Q:    0.0586  : Inter-comparmental CL (L/h)
TV_V2:   3.05    : Volume of peripheral compartment (L)
ARV:  2.07    : Additive RUV
PRV:  0.23    : Proportional RUV


[OMEGA] @annotated
Eta_CL: 0.604   : ETA on clearance
Eta_V1: 0.232   : ETA on central volume
Eta_V2: 0.738   : ETA on peripheral volume


[SIGMA] @labels EPS1 EPS2
1 1


[CMT] @annotated
CENT:  central cmt (ug)
PERI:  peripheral cmt (ug)


[MAIN]
double CL = TV_CL *exp(Eta_CL);
double V1 = TV_V1 *exp(Eta_V1);
double Q  = TV_Q;
double V2 = TV_V2 *exp(Eta_V2);


[ODE]
dxdt_CENT  = -(CL/V1)*CENT - (Q/V1)*CENT + (Q/V2)*PERI;
dxdt_PERI  =  (Q/V1)*CENT - (Q/V2)*PERI;


[TABLE]
double C_CENT = CENT/V1;
double C_OBS = C_CENT*(1+PRV*EPS1) + ARV*EPS2;


[CAPTURE] @annotated
C_CENT  : Central CMT conc (ug/mL)
C_OBS   : Central CMT conc with error (ug/mL)


[SET]
end = 4200, delta = 2, digits = 4
"
```

![image](https://user-images.githubusercontent.com/29249244/169718412-2f747c58-ca8a-4275-999a-421fc3393bba.png)

**Comment by kylebaron** [comment]
Hi @mark0935git  - 

One way to handle this is to re-simulate `EPS` until `C_OBS` is not negative; this uses the `simeta()` function to draw a new set of `EPS()`; we wrap this in a `while` statement and keep a counter to make sure that we don't go into some infinite loop

```r
[TABLE]
double C_CENT = CENT/V1;
double C_OBS = C_CENT*(1+PRV*EPS1) + ARV*EPS2;

int i = 0;
while(C_OBS < 0 && i++ < 20) {
  simeps();
  C_OBS = C_CENT*(1+PRV*EPS1) + ARV*EPS2;
}
```

I'm not sure if the simulated `C_OBS` of  `-3.662` are pre- first dose or not; might just force those to zero if `CENT` is zero.  But if `C_OBS` is getting smaller at the end of the dosing interval and you're simulating negative concentrations, you might have to have the `simeps()` code anyway. 

I think this is the way to handle it if you want to keep the existing error model. The exponential error model would be equivalent to the proportional error; there also is a way to  the equivalent of additive + proportional error on log-DV but that would change the model. 

Kyle

**Comment by smouksassi** [comment]
depending on what you are doing forcing the C_Obs to be > 0 via keep simulating  until I find an epsilon I like,
instead of censoring values below a quantification limit might be in effect using a different data generation process from the one we assumed we collected the data from as Kyle mentioned ideally you use a  transformation + error model that cannot simulate illegal DV values.
e.g.:
https://www.diva-portal.org/smash/get/diva2:941121/FULLTEXT01.pdf

and link to the classic Beal double exp approach:
https://nonmem.iconplc.com/#/nonmem/tips --> tip 9
$ERROR
PRD=F ;individual non-transformed prediction (with POSTHOC or FOCE)
;otherwise, the population non-transformed prediction
;Beal, JPP 28(5), 2001, p.488
Y=LOG((F)+THETA(3))+(F*ERR(1))/(F+THETA(3))+(THETA(3)*ERR(2))/(F+THETA(3)) 
IF(F.GT.0) THEN
IPRED=LOG(F) ;the log-transformed individual prediction
ELSE
IPRED=0
ENDIF
$THETA
   (0, 0.1) ;"m, a positively constrained variance parameter" theta3

**Comment by mark0935git** [comment]
@smouksassi @kylebaron Thanks for your explanation! The problem is resolved nicely.
